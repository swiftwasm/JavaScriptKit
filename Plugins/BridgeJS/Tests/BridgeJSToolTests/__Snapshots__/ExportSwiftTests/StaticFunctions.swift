// NOTICE: This is auto-generated code by BridgeJS from JavaScriptKit,
// DO NOT EDIT.
//
// To update this file, just rebuild your project or run
// `swift package bridge-js`.

@_spi(BridgeJS) import JavaScriptKit

extension Calculator: _BridgedSwiftCaseEnum {
    @_spi(BridgeJS) @_transparent public consuming func bridgeJSLowerParameter() -> Int32 {
        return bridgeJSRawValue
    }
    @_spi(BridgeJS) @_transparent public static func bridgeJSLiftReturn(_ value: Int32) -> Calculator {
        return Calculator(bridgeJSRawValue: value)!
    }
    @_spi(BridgeJS) @_transparent public static func bridgeJSLiftParameter(_ value: Int32) -> Calculator {
        return Calculator(bridgeJSRawValue: value)!
    }
    @_spi(BridgeJS) @_transparent public consuming func bridgeJSLowerReturn() -> Int32 {
        return bridgeJSRawValue
    }

    private init?(bridgeJSRawValue: Int32) {
        switch bridgeJSRawValue {
        case 0:
            self = .scientific
        case 1:
            self = .basic
        default:
            return nil
        }
    }

    private var bridgeJSRawValue: Int32 {
        switch self {
        case .scientific:
            return 0
        case .basic:
            return 1
        }
    }
}

@_expose(wasm, "bjs_Calculator_static_square")
@_cdecl("bjs_Calculator_static_square")
public func _bjs_Calculator_static_square(value: Int32) -> Int32 {
    #if arch(wasm32)
    let ret = Calculator.square(value: Int.bridgeJSLiftParameter(value))
    return ret.bridgeJSLowerReturn()
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

extension APIResult: _BridgedSwiftAssociatedValueEnum {
    @_spi(BridgeJS) @_transparent public static func bridgeJSLiftParameter(_ caseId: Int32) -> APIResult {
        switch caseId {
        case 0:
            return .success(String.bridgeJSLiftParameter(_swift_js_pop_param_int32(), _swift_js_pop_param_int32()))
        case 1:
            return .failure(Int.bridgeJSLiftParameter(_swift_js_pop_param_int32()))
        default:
            fatalError("Unknown APIResult case ID: \(caseId)")
        }
    }

    @_spi(BridgeJS) @_transparent public consuming func bridgeJSLowerReturn() {
        switch self {
        case .success(let param0):
            _swift_js_push_tag(Int32(0))
            var __bjs_param0 = param0
            __bjs_param0.withUTF8 { ptr in
                _swift_js_push_string(ptr.baseAddress, Int32(ptr.count))
            }
        case .failure(let param0):
            _swift_js_push_tag(Int32(1))
            _swift_js_push_int(Int32(param0))
        }
    }
}

@_expose(wasm, "bjs_APIResult_static_roundtrip")
@_cdecl("bjs_APIResult_static_roundtrip")
public func _bjs_APIResult_static_roundtrip(value: Int32) -> Void {
    #if arch(wasm32)
    let ret = APIResult.roundtrip(value: APIResult.bridgeJSLiftParameter(value))
    return ret.bridgeJSLowerReturn()
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

@_expose(wasm, "bjs_Utils_String_uppercase")
@_cdecl("bjs_Utils_String_uppercase")
public func _bjs_Utils_String_uppercase(textBytes: Int32, textLength: Int32) -> Void {
    #if arch(wasm32)
    let ret = Utils.String.uppercase(_: String.bridgeJSLiftParameter(textBytes, textLength))
    return ret.bridgeJSLowerReturn()
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

@_expose(wasm, "bjs_MathUtils_init")
@_cdecl("bjs_MathUtils_init")
public func _bjs_MathUtils_init() -> UnsafeMutableRawPointer {
    #if arch(wasm32)
    let ret = MathUtils()
    return ret.bridgeJSLowerReturn()
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

@_expose(wasm, "bjs_MathUtils_static_subtract")
@_cdecl("bjs_MathUtils_static_subtract")
public func _bjs_MathUtils_static_subtract(a: Int32, b: Int32) -> Int32 {
    #if arch(wasm32)
    let ret = MathUtils.subtract(a: Int.bridgeJSLiftParameter(a), b: Int.bridgeJSLiftParameter(b))
    return ret.bridgeJSLowerReturn()
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

@_expose(wasm, "bjs_MathUtils_static_add")
@_cdecl("bjs_MathUtils_static_add")
public func _bjs_MathUtils_static_add(a: Int32, b: Int32) -> Int32 {
    #if arch(wasm32)
    let ret = MathUtils.add(a: Int.bridgeJSLiftParameter(a), b: Int.bridgeJSLiftParameter(b))
    return ret.bridgeJSLowerReturn()
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

@_expose(wasm, "bjs_MathUtils_multiply")
@_cdecl("bjs_MathUtils_multiply")
public func _bjs_MathUtils_multiply(_self: UnsafeMutableRawPointer, x: Int32, y: Int32) -> Int32 {
    #if arch(wasm32)
    let ret = MathUtils.bridgeJSLiftParameter(_self).multiply(x: Int.bridgeJSLiftParameter(x), y: Int.bridgeJSLiftParameter(y))
    return ret.bridgeJSLowerReturn()
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

@_expose(wasm, "bjs_MathUtils_deinit")
@_cdecl("bjs_MathUtils_deinit")
public func _bjs_MathUtils_deinit(pointer: UnsafeMutableRawPointer) {
    Unmanaged<MathUtils>.fromOpaque(pointer).release()
}

extension MathUtils: ConvertibleToJSValue, _BridgedSwiftHeapObject {
    var jsValue: JSValue {
        #if arch(wasm32)
        @_extern(wasm, module: "TestModule", name: "bjs_MathUtils_wrap")
        func _bjs_MathUtils_wrap(_: UnsafeMutableRawPointer) -> Int32
        #else
        func _bjs_MathUtils_wrap(_: UnsafeMutableRawPointer) -> Int32 {
            fatalError("Only available on WebAssembly")
        }
        #endif
        return .object(JSObject(id: UInt32(bitPattern: _bjs_MathUtils_wrap(Unmanaged.passRetained(self).toOpaque()))))
    }
}