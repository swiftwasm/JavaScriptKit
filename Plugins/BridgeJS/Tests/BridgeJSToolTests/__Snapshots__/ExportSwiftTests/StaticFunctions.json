{
  "classes" : [
    {
      "constructor" : {
        "abiName" : "bjs_MathUtils_init",
        "effects" : {
          "isAsync" : false,
          "isStatic" : false,
          "isThrows" : false
        },
        "parameters" : [

        ]
      },
      "methods" : [
        {
          "abiName" : "bjs_MathUtils_static_subtract",
          "effects" : {
            "isAsync" : false,
            "isStatic" : true,
            "isThrows" : false
          },
          "name" : "subtract",
          "parameters" : [
            {
              "label" : "a",
              "name" : "a",
              "type" : {
                "int" : {

                }
              }
            },
            {
              "label" : "b",
              "name" : "b",
              "type" : {
                "int" : {

                }
              }
            }
          ],
          "returnType" : {
            "int" : {

            }
          },
          "staticContext" : {
            "className" : {
              "_0" : "MathUtils"
            }
          }
        },
        {
          "abiName" : "bjs_MathUtils_static_add",
          "effects" : {
            "isAsync" : false,
            "isStatic" : true,
            "isThrows" : false
          },
          "name" : "add",
          "parameters" : [
            {
              "label" : "a",
              "name" : "a",
              "type" : {
                "int" : {

                }
              }
            },
            {
              "label" : "b",
              "name" : "b",
              "type" : {
                "int" : {

                }
              }
            }
          ],
          "returnType" : {
            "int" : {

            }
          },
          "staticContext" : {
            "className" : {
              "_0" : "MathUtils"
            }
          }
        },
        {
          "abiName" : "bjs_MathUtils_multiply",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "multiply",
          "parameters" : [
            {
              "label" : "x",
              "name" : "x",
              "type" : {
                "int" : {

                }
              }
            },
            {
              "label" : "y",
              "name" : "y",
              "type" : {
                "int" : {

                }
              }
            }
          ],
          "returnType" : {
            "int" : {

            }
          }
        }
      ],
      "name" : "MathUtils",
      "properties" : [

      ],
      "swiftCallName" : "MathUtils"
    }
  ],
  "enums" : [
    {
      "cases" : [
        {
          "associatedValues" : [

          ],
          "name" : "scientific"
        },
        {
          "associatedValues" : [

          ],
          "name" : "basic"
        }
      ],
      "emitStyle" : "const",
      "name" : "Calculator",
      "staticMethods" : [
        {
          "abiName" : "bjs_Calculator_static_square",
          "effects" : {
            "isAsync" : false,
            "isStatic" : true,
            "isThrows" : false
          },
          "name" : "square",
          "parameters" : [
            {
              "label" : "value",
              "name" : "value",
              "type" : {
                "int" : {

                }
              }
            }
          ],
          "returnType" : {
            "int" : {

            }
          },
          "staticContext" : {
            "enumName" : {
              "_0" : "Calculator"
            }
          }
        }
      ],
      "staticProperties" : [

      ],
      "swiftCallName" : "Calculator"
    },
    {
      "cases" : [
        {
          "associatedValues" : [
            {
              "type" : {
                "string" : {

                }
              }
            }
          ],
          "name" : "success"
        },
        {
          "associatedValues" : [
            {
              "type" : {
                "int" : {

                }
              }
            }
          ],
          "name" : "failure"
        }
      ],
      "emitStyle" : "const",
      "name" : "APIResult",
      "staticMethods" : [
        {
          "abiName" : "bjs_APIResult_static_roundtrip",
          "effects" : {
            "isAsync" : false,
            "isStatic" : true,
            "isThrows" : false
          },
          "name" : "roundtrip",
          "parameters" : [
            {
              "label" : "value",
              "name" : "value",
              "type" : {
                "associatedValueEnum" : {
                  "_0" : "APIResult"
                }
              }
            }
          ],
          "returnType" : {
            "associatedValueEnum" : {
              "_0" : "APIResult"
            }
          },
          "staticContext" : {
            "enumName" : {
              "_0" : "APIResult"
            }
          }
        }
      ],
      "staticProperties" : [

      ],
      "swiftCallName" : "APIResult"
    },
    {
      "cases" : [

      ],
      "emitStyle" : "const",
      "name" : "String",
      "namespace" : [
        "Utils"
      ],
      "staticMethods" : [
        {
          "abiName" : "bjs_Utils_String_uppercase",
          "effects" : {
            "isAsync" : false,
            "isStatic" : true,
            "isThrows" : false
          },
          "name" : "uppercase",
          "parameters" : [
            {
              "label" : "_",
              "name" : "text",
              "type" : {
                "string" : {

                }
              }
            }
          ],
          "returnType" : {
            "string" : {

            }
          },
          "staticContext" : {
            "namespaceEnum" : {
              "_0" : "Utils.String"
            }
          }
        }
      ],
      "staticProperties" : [

      ],
      "swiftCallName" : "Utils.String"
    },
    {
      "cases" : [

      ],
      "emitStyle" : "const",
      "name" : "Utils",
      "staticMethods" : [

      ],
      "staticProperties" : [

      ],
      "swiftCallName" : "Utils"
    }
  ],
  "functions" : [

  ],
  "moduleName" : "TestModule"
}